nschecker/notify.go

```diff
@@ -0,0 +1,99 @@
+package nschecker
+
+import (
+	"fmt"
+	"log"
+	"sync"
+
+	"net/http"
+	"net/url"
+
+	"strings"
+
+	"github.com/nlopes/slack"
+)
+
+// Notifier interface is construct an response.
+type Notifier interface {
+	Notify(state State, s Source) error
+}
+
+// SlackNotifier struct is construct an slack message.
+type SlackNotifier struct {
+	Cli *slack.Client
+
+	channel string
+
+	// url -> current state
+	statesMu sync.Mutex
+	states   map[string]State
+}
+
+func NewSlackNotifier(cli *slack.Client, channel string) Notifier {
+	return &SlackNotifier{
+		Cli:     cli,
+		channel: channel,
+		states:  make(map[string]State),
+	}
+}
+
+func (n *SlackNotifier) Notify(state State, s Source) error {
+	defer func() {
+		n.statesMu.Lock()
+		n.states[s.URL] = state
+		n.statesMu.Unlock()
+	}()
+	n.statesMu.Lock()
+	oldState, ok := n.states[s.URL]
+	n.statesMu.Unlock()
+	if !ok && state == SOLDOUT {
+		return nil
+	}
+	if oldState == state {
+		log.Printf("same state: %v url=%v name=%v", state, s.URL, s.Name)
+		return nil
+	}
+	channel := ""
+	if state == AVAILABLE {
+		channel = "<!channel|channel> "
+	}
+	msg := fmt.Sprintf("%s%v: %v (%v)", channel, state, s.URL, s.Name)
+	params := slack.PostMessageParameters{EscapeText: false}
+	_, _, err := n.Cli.PostMessage(n.channel, msg, params)
+	return err
+}
+
+// LineNotifier struct is construct an LINE message.
+type LineNotifier struct {
+	hc  *http.Client
+	tok string
+}
+
+func NewLineNotifier(hc *http.Client, token string) Notifier {
+	return &LineNotifier{
+		hc:  hc,
+		tok: token,
+	}
+}
+
+func (n *LineNotifier) Notify(state State, s Source) error {
```

Comment:
Can you make this logic same as slack notifier?
I mean, we shouldn't notify status unless the state is changed to available.

Current implementation is unusable because it notifies sold out state in every interval.